tools/cldr-code/src/main/java/org/unicode/cldr/tool/GenerateDisplayNameTestData.java

```diff
@@ -0,0 +1,121 @@
+import org.unicode.cldr.util.CLDRPaths;
+import org.unicode.cldr.util.ChainedMap;
+import org.unicode.cldr.util.ChainedMap.M3;
+import org.unicode.cldr.util.CldrUtility;
+import org.unicode.cldr.util.Factory;
+import org.unicode.cldr.util.TempPrintWriter;
+import org.unicode.cldr.util.XPathParts;
+
+public class GenerateDisplayNameTestData {
+
+  static CLDRConfig CLDR_CONFIG = CLDRConfig.getInstance();
+
+  static final String LANGUAGE_PATH = "//ldml/localeDisplayNames/languages/language";
+
+  /**
+   *
+   * @param args
+   */
+  public static void main(String[] args) {
+    // Parse CLDR data files to get specific data of interest for creating test cases
+    M3<String, String, String> testCases = getTestCases();
+
+    try (TempPrintWriter pw =
+        TempPrintWriter.openUTF8Writer(
+            CLDRPaths.TEST_DATA + "localeIdentifiers", "displayNames.txt")) {
+
+      pw.println(
+          "# Test data for Display Names\n"
+              + CldrUtility.getCopyrightString("#  ")
+              + "\n"
+              + "#\n"
+              + "# Test data for https://unicode.org/reports/tr35/tr35-general.html#Display_Name_Elements\n"
+              + "#\n"
+              + "# Format:\n"
+              + "# Formatting locale\tInput locale\tFormatted output\n");
+
+      for (Entry<String, Map<String, String>> entry : testCases) {
+        String formattingLocale = entry.getKey();
+        for (Entry<String, String> entry2 : entry.getValue().entrySet()) {
+          String inputLocale = entry2.getKey();
+          String output = entry2.getValue();
+
+          pw.println(formattingLocale + "\t" + inputLocale + "\t" + output);
+        }
+      }
+
+    }
+  }
+
+  public static M3<String, String, String> getTestCases() {
+    // first-level key (K2 type parameter in class M3): formatting locale
+    // second-level key (K1 type parameter in class M3): input locale for translation
+    // value (V type parameter in class M3): formatted locale string
+    M3<String, String, String> displayNames =
+        ChainedMap.of(
+            new TreeMap<String, Object>(),
+            new TreeMap<String, Object>(),
+            String.class);
+
+    Factory factory = CLDR_CONFIG.getCldrFactory();
+
+    ULocale undLocale = new ULocale("und");
+
+    Pattern removeLetters = Pattern.compile("\\P{Alnum}");
+
+    main:
+    for (String locale : factory.getAvailableLanguages()) {
+
+      try {
+        CLDRFile cldrFile =
+            factory.make(locale, true, DraftStatus.contributed); // don't include
+        // draft=unconfirmed/provisional
+        CLDRFile unresolved = cldrFile.getUnresolved();
+
+        // Check that we have sufficient display name data for this locale, else skip
+
+        for (String path : cldrFile) {
+          // only when we match do we get & process the data at that path
+          if (path.startsWith(LANGUAGE_PATH)) {
+            // Get display name
+            String value = cldrFile.getStringValue(path);
```

Comment:
Each cldrFile has a method to create a locale name (optionally with language, script, and region). It would be better to use that.

Clearly we don't want to produce all the combinations. 

I would suggest. Take the locales in Organization.cldr (call that LOC)
for locale in LOC
  for localeForName in LOC
      test to see if each localeForName's pieces is in Organization.cldr's coverage for locale
      skip if not
      emit the locale identifier and the name of localeForName in locale

An enhancement of that would be to also have alt values.

Reply:
After more work with @sffc , we pared down the generated test data to a small list of input locales that exercise interesting corner cases (lower & higher coverage levels, which pieces of LSR are used to form the dialect display name, etc.).

We also moved the logic into a preexisting test data file generator (rather than creating a new class file of code and output file with a duplicate purpose).

We went ahead and ensure that both the {standard, dialect} styles of display are used to generate test cases, whereas it previously only used standard.